### DISCUSSION TITLE
Tab contention?

### DISCUSSION THREAD

**Discussion by lukasb** [dexie/Dexie.js#2107] [Q&A]
I'm encountering an intermittent issue in which useLiveQuery will sometimes return undefined despite IndexedDB having data.

- Only occurs on mobile Webkit (maybe desktop but haven't tried hard to reproduce here)
- No error is thrown in useLiveQuery
- Only happens when other tabs are also open to my app
- When this happens, opening another tab to my app will always exhibit the same problem
- Closing all other tabs open to my app instantly fixes the problem, useLiveQuery then successfully returns data. (Usually don't have to close all other tabs to get it to work, just closing some is enough.)

It's almost as if one of the tabs is stuck in a state where it's keeping a lock that prevents the other tabs from reading the db, and closing that tab fixes the problem. But I don't have any code in my app that would prevent other tabs from reading from the db.

```
const pages = useLiveQuery(async () => {

    if (!session || !session.id) return [];

    const localPages = await localDb.pages
      .where("userId")
      .equals(session.id)
      .toArray();

    const queuedUpdates = await localDb.queuedUpdates
      .where("userId")
      .equals(session.id)
      .toArray();

    if (!localPages || !queuedUpdates) return undefined;

    const mergedPages = [
      ...localPages
        // remove deleted pages
        .filter((page) => !page.deleted)
        // remove pages with queued updates marking them as deleted
        .filter((page) => {
          const queuedUpdate = queuedUpdates.find(
            (update) => update.id === page.id
          );
          return !queuedUpdate?.deleted;
        })
        // replace with queued updates if they exist
        .map((page) => {
          const queuedUpdate = queuedUpdates.find(
            (update) => update.id === page.id
          );
          return queuedUpdate || page;
        }),
      // add queued updates that aren't in the main table
      ...queuedUpdates.filter(
        (update) =>
          !localPages.some((page) => page.id === update.id) && !update.deleted
      ),
    ];
    return mergedPages;
  }, [session]);
  ```

**Comment by dfahlander** [comment]
It is normal that useLiveQuery() returns *undefined* on initial render before the async querier is done with its result. Is this the case or do you see it returning undefined also after initial render?

**Reply by lukasb** [reply]
Also after initial render. I have a component on the page that displays "pages undefined" if pages is undefined, and 99% of the time I see a flash of that before content displays. 1% of the time it stays stuck there.

**Reply by dfahlander** [reply]
I'm wondering if this could be related to the bfcache support introduced in dexie@4.0.10. Are you seeing the same issue if you try setting `Dexie.disableBfCache = true` somewhere where your Dexie instance is created? If that would make a difference, it could be something about how webkit fires the pagehide and pageshow events.

**Reply by lukasb** [reply]
I just checked and I'm using 4.0.9 :/

**Reply by dfahlander** [reply]
Try if 4.0.10 solves it. There was a bug fixed in 4.0.10: https://github.com/dexie/Dexie.js/commit/df9b6f0693b65632f496c5c4d58b7c68a499dbf8 . The bfcache support was actually introduced in 4.0.1 (not 4.0.10 as I wrote in last comment) but there was a fix in 4.0.10 for it.

**Reply by lukasb** [reply]
Upgraded to 4.0.10 and added disableBfCache (see below) but it's not fixed yet.

Also, when this state gets triggered (with one tab getting only undefined from useLiveQuery) other tabs that have already loaded content from useLiveQuery are no longer able to write to IndexedDB.

Will see what else I can figure out.

```
Dexie.disableBfCache = true;

const localDb = new Dexie('orangetask-local', { cache: 'immutable' }) as Dexie & {
  pages: EntityTable<Page, 'id'>;
  queuedUpdates: EntityTable<Page, 'id'>;
  lastRevisionSynced: EntityTable<{ id: string, revisionNumber: number }, 'id'>;
};
```

**Reply by dfahlander** [reply]
Ok, seems that it could be something that keeps the database locked som somehow on a lower level. Since you mention webkit mobile, I suppose it's an iOS? Whats version is it?

**Reply by lukasb** [reply]
iOS 18.1.1

**Comment by lukasb** [comment]
Some more information if it's helpful ...

useLiveQuery only gets called once. when it gets called ...

- session.id is defined
- localDb.isOpen() returns false
- localDb.pages and localDb.queuedUpdates are defined
- no code is executed after await localDb.pages.where("userId") etc - no error is thrown either

I have a scheduled task in my app that runs every 30 seconds, when that happens ...

- localDb.isOpen() returns true
- same thing, no code is executed after await localDb.pages.where("id") etc - and again, no error is thrown

**Comment by lukasb** [comment]
FWIW have not seen this in about a month
